hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/commit/integration/ITestS3ACommitterMRJob.java

```diff
@@ -0,0 +1,643 @@
+import org.junit.runners.Parameterized;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.s3a.S3AFileSystem;
+import org.apache.hadoop.fs.s3a.S3AUtils;
+import org.apache.hadoop.fs.s3a.commit.AbstractYarnClusterITest;
+import org.apache.hadoop.fs.s3a.commit.CommitConstants;
+import org.apache.hadoop.fs.s3a.commit.LoggingTextOutputFormat;
+import org.apache.hadoop.fs.s3a.commit.files.SuccessData;
+import org.apache.hadoop.fs.s3a.commit.magic.MagicS3GuardCommitter;
+import org.apache.hadoop.fs.s3a.commit.staging.DirectoryStagingCommitter;
+import org.apache.hadoop.fs.s3a.commit.staging.PartitionedStagingCommitter;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.mapreduce.JobStatus;
+import org.apache.hadoop.mapreduce.Mapper;
+import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
+import org.apache.hadoop.mapreduce.lib.input.TextInputFormat;
+import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hadoop.util.DurationInfo;
+
+import static org.apache.hadoop.fs.s3a.S3ATestUtils.disableFilesystemCaching;
+import static org.apache.hadoop.fs.s3a.S3ATestUtils.lsR;
+import static org.apache.hadoop.fs.s3a.S3AUtils.applyLocatedFiles;
+import static org.apache.hadoop.fs.s3a.commit.CommitConstants.FS_S3A_COMMITTER_STAGING_TMP_PATH;
+import static org.apache.hadoop.fs.s3a.commit.CommitConstants.MAGIC;
+import static org.apache.hadoop.fs.s3a.commit.CommitConstants._SUCCESS;
+import static org.apache.hadoop.fs.s3a.commit.InternalCommitterConstants.FS_S3A_COMMITTER_STAGING_UUID;
+import static org.apache.hadoop.fs.s3a.commit.staging.Paths.getMultipartUploadCommitsDirectory;
+import static org.apache.hadoop.fs.s3a.commit.staging.StagingCommitterConstants.STAGING_UPLOADS;
+import static org.apache.hadoop.test.LambdaTestUtils.intercept;
+
+/**
+ * Test an MR Job with all the different committers.
+ * <p>
+ * This is a fairly complex parameterization: it is designed to
+ * avoid the overhead of starting a Yarn cluster for
+ * individual committer types, so speed up operations.
+ * <p>
+ * It also implicitly guarantees that there is never more than one of these
+ * MR jobs active at a time, so avoids overloading the test machine with too
+ * many processes.
+ * How the binding works:
+ * <ol>
+ *   <li>
+ *     Each parameterized suite is configured through its own
+ *     {@link CommitterTestBinding} subclass.
+ *   </li>
+ *   <li>
+ *     JUnit runs these test suites one parameterized binding at a time.
+ *   </li>
+ *   <li>
+ *     The test suites are declared to be executed in ascending order, so
```

Comment:
A little confused by the intent of this flow - i.e. test_000, test_100, test_200, test_500 in order.
1) What happens if test_000 fails - will the rest of the tests still be run for the specific parameter?
2) The names - test_000, test_100 etc don't say a lot about what is actually being done in the methods.

It may be easier to have a single test method - which takes care of setup, validation, etc - and allow override points within that for the individual committers. That would lead to a single test failing and not attempting to run the others after a previous dependent failure.

Also, mentioned making the cluster non-static in another comment. Given parameterized tests instantiate a new instance of the class each time - maybe just the '@AfterClass' teerdown and cluster setup needs to move to this class (still as static).

Reply:
test suite is tagged as `@FixMethodOrder(MethodSorters.NAME_ASCENDING)`; they run in sequence. naming convention is as used elsewhere. 

As to why the gaps between numbers: lets us put new values in without
renaming things. You never coded in Basic or APL?

Having separate tests 
* isolates test method timeouts
* lets you debug individual operations.
* makes it clear which operation failed.
yes, the state of one is dependent  on the others succeeding, which is why they try to have checks for the state of the store before they run.

Reply:
thinking: we could set markers on tests as they complete and so skip the successors. I'll do that

Reply:
Not concerned about the gap in numbering at all.

I understand the ordering enforced by the annotation. The terrasort test has this as well - and it makes sense to me there, at least to some extent, since each test is doing something independently, which can be verified (e.g. run teragen, then terasort, then teravalidate).

In this case - some of the methods are just performing setup operations for the actual test.

This could be structured as.
```
@Test testMRJob() {
   committerTestBinding.test_000();
   committerTestBinding.test_100();
   .. .. ..
}
```
That would make sure exactly one test runs, and fails the moment there's an error from any of the per committer bindings. Is also easier to debug I think - for someone new, if they see a failure on test_000, test_100, test_500 etc - they would not try to figure out each of the failures independently. Will also have context on the failures via the methodName + parameter.

IAC, that's my take on this. Have a preference for a single test approach, but not tied to it. Your plan to skip any successors solves the main concern as well.

Reply:
I am satisfied with what we have here -it is consistent with all the other large sequential test runs

Reply:
Sounds good to me. The change that you mentioned about not running tests if a previous test fails - if that's already in, I'm +1 for the patch.